### DISCUSSION TITLE
how to save a pdf page to png bytes with typing [pyright issue]

### DISCUSSION THREAD

**Discussion by jaredmcqueen** [pymupdf/PyMuPDF#3849] [Looking for help] [answered]
this is how i'm currently saving a pdf page to png bytes:

    page: Page = doc[page_num]
    log.info(f"Processing page {page_num}")

    pix = page.get_pixmap()
    img = Image.frombytes("RGB", [pix.width, pix.height], pix.samples)

    png_bytes = io.BytesIO()

    img.save(png_bytes, format="PNG")
    png_bytes = png_bytes.getvalue()

it works, but there is zero typing, and pyright throws a fit with every step of the way.

 error: Cannot access attribute "get_pixmap" for class "Page"
    Attribute "get_pixmap" is unknown (reportAttributeAccessIssue)
    
also for the `pix.width` and `pix.height`:

Argument of type "list[Unknown]" cannot be assigned to parameter "size" of type "tuple[int, int]" in function "frombytes"
    "list[Unknown]" is not assignable to "tuple[int, int]" (reportArgumentType)

**Comment by JorjMcKie** [comment]
Can you explain why you are doing this in such a needlessly complicated way? Two statements are enough!
```python
pix = page.get_pixmap()
pix.save("some.png")
```

**Comment by jaredmcqueen** [comment]
I need the png in bytes to send to a vision model.  I didn't want to save to disk, just to re-open it as bytes.  I'm processing a lot of pages and didn't want to thrash the disk with lots of IO when I could just keep the data in memory.

I'm still learning the nuances of this library so there's probably a better way to get the png version of each page in bytes, but so far I can't find a better way.  Thanks for the help!

**Reply by JorjMcKie** [reply, marked as answer]
"Saving" to memory is just as simple:
```python
pix = page.get_text()
pngbytes = pix.tobytes("png")
```
Looking in help():
```python
import pymupdf
help(pymupdf.Pixmap.tobytes)
Help on function tobytes in module pymupdf:

tobytes(self, output='png', jpg_quality=95)
    Convert to binary image stream of desired type.
```

**Reply by JorjMcKie** [reply]
And here is the [documentation](https://pymupdf.readthedocs.io/en/latest/pixmap.html#Pixmap.tobytes).
So where did you look instead 🤷‍♂️😉?

**Reply by jaredmcqueen** [reply]
good call and thanks for pointing me in the right direction.

I tend to rely too heavily on "going to definition" to read source code of libraries, which doesn't seem to work with this library.  I wrongly assumed that since type hints or autocomplete didn't work that I was doing something wrong.

<img width="316" alt="image" src="https://github.com/user-attachments/assets/933324d4-249e-4389-bd42-ca2076da2981">

**Reply by JorjMcKie** [reply]
PyMuPDF's documentation in PDF format is over 550 pages in size and of the best of its kind. Also includes several dozen examples.
It also makes perfect sense to look around in this repo: https://github.com/pymupdf/PyMuPDF-Utilities

**Comment by jaredmcqueen** [comment]
closing
